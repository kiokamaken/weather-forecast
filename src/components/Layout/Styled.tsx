import styled from 'styled-components';
import { Layout } from 'antd';

export const MainContentStyled = styled(Layout.Content)`
  width: 85%;
  max-width: 1200px;
  margin: auto;
`;
