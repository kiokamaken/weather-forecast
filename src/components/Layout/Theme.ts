import { DefaultTheme } from 'styled-components'

export const base: DefaultTheme = {
  palette: {
    primary: '#ed0cef',
    secondary: '#8105d8',
    white: '#fff',
    black: '#000',
    grey: '#ecf0f1',
    background: '#3b064d'
  },
  shadow: '0px 3px 15px 2px #ddd',
  animation: 'all 0.3s ease-in-out',
  fontSizes: {
    small: '0.75rem',
    normal: '1rem',
    medium: '1.5rem',
    large: '2rem',
  },
};

export const dark = { ...base };
export const light = { ...base };
