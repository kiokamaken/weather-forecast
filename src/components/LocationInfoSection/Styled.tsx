import styled from 'styled-components';

export const LocationWrapper = styled.article`
  padding: 1rem 2rem;
`;
