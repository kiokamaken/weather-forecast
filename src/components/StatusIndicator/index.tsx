import React from 'react';
import { Spin } from 'antd';
import {
  LoadingOutlined,
  CheckCircleTwoTone,
  ExclamationCircleTwoTone,
} from '@ant-design/icons';

export interface StatusIndicatorProps {
  loading: boolean;
  success?: boolean;
  warning?: boolean;
}

const StatusIndicator = (props: StatusIndicatorProps) => {
  const { loading, success, warning } = props;
  return (
    <>
      {loading && (
        <Spin indicator={<LoadingOutlined style={{ fontSize: 24 }} spin />} />
      )}
      {!loading && success && <CheckCircleTwoTone style={{ fontSize: 24 }} />}
      {!loading && warning && (
        <ExclamationCircleTwoTone twoToneColor="red" style={{ fontSize: 24 }} />
      )}
    </>
  );
};

export default React.memo(StatusIndicator);
