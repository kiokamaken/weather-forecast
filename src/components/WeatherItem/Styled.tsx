import styled from 'styled-components';

export const StyledWeatherItem = styled.section`
  border-radius: 0.5rem;
  padding: 1rem 0.5rem 0.5rem;
  background: ${(props) => props.theme.palette.white};
  position: relative;
  margin-top: 4rem;

  &:before {
    content: '';
    display: block;
    width: 6rem;
    height: 6rem;
    background: ${(props) => props.theme.palette.white};
    border-radius: 50%;
    position: absolute;
    transform: translate(0%,-65%);
    left: 0;
    right: 0;
    margin: auto;
    z-index: -1;
  }

  > img {
    width: 4.5rem;
    height: auto;
    display: block;
    margin: 0.5rem auto;
    position: absolute;
    transform: translate(0%,-82%);
    left: 0;
    right: 0;
  }

  .day {
    display: block;
    text-align: center;
    padding: 0rem 0.5rem;
    margin-top: 2rem;
  }

  .the-temp {
    text-align: center;
    margin-bottom: 0.5rem;
  }
`;
