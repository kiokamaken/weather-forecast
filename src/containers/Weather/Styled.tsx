import styled from 'styled-components';
import headerImage from 'header-image.jpg';

export const PageHeader = styled.header`
  background-image: url(${headerImage});
  background-position: center center;
  background-size: cover;
  margin-bottom: 1rem;
  height: 10rem;
  border-bottom-left-radius: 2rem;
  border-bottom-right-radius: 2rem;
`;
