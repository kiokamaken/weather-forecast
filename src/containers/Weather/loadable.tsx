import React from 'react';
import loadable from '@loadable/component';

const LoadableComponent = loadable(() => import('./index'));

const LoadableWeather = () => <LoadableComponent />;

export default LoadableWeather;
