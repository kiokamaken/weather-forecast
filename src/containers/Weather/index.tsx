import { AutoComplete, Col, message, Select, Switch } from 'antd';
import MainContent from 'components/Layout/MainContent';
import Row from 'components/Layout/Row';
import LocationInfoSection from 'components/LocationInfoSection';
import StatusIndicator from 'components/StatusIndicator';
import WeatherItem from 'components/WeatherItem';
import debounce from 'lodash/fp/debounce';
import { Location, LocationBase } from 'models';
import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  selectError,
  selectForecastData,
  selectLoading,
  selectLocationSuggestion,
} from 'store/weather/selectors';
import {
  getForecastByWoeId,
  getLocationSuggestions,
} from 'store/weather/slice';

import { PageHeader } from './Styled';

const { Option } = Select;

export interface WeatherProps {
  options: LocationBase[];
  forecastData: Location;
  loading: boolean;
  error: string[];
  getForecastByWoeId: Function;
  getLocationSuggestions: Function;
}

export class Weather extends React.Component<WeatherProps> {
  state = {
    isCelsius: true,
  };

  searchType = 'location';

  componentDidUpdate(prevProp: WeatherProps) {
    if (this.props.error?.length && !prevProp.error?.length) {
      message.error(this.props.error[0]);
    }
  }

  onSearch = debounce(
    300,
    (search: string) =>
      search.trim() !== '' &&
      this.props.getLocationSuggestions({
        search,
        isLattLong: this.searchType === 'lattlong',
      })
  );

  onSelect = (value: string, option: any) =>
    value && this.props.getForecastByWoeId({ id: option.key });

  onChangeUnit = () => this.setState({ isCelsius: !this.state.isCelsius });

  onChangeSearchType = (value: string) => (this.searchType = value);

  render() {
    const { forecastData, options, loading, error } = this.props;
    return (
      <MainContent>
        <PageHeader>
          {forecastData && <LocationInfoSection info={forecastData} />}
        </PageHeader>
        <Row align="middle">
          <Col xs={{ order: 1, span: 10 }} sm={4}>
            <Select
              defaultValue={'location'}
              onChange={this.onChangeSearchType}
            >
              <Option value="location">Location</Option>
              <Option value="lattlong">LattLong</Option>
            </Select>
          </Col>
          <Col xs={{ order: 4, span: 22 }} sm={{ order: 2, span: 6 }}>
            <AutoComplete
              onSelect={this.onSelect}
              onSearch={this.onSearch}
              placeholder="Input Location here"
            >
              {options &&
                options.map((option) => (
                  <AutoComplete.Option key={option.woeid} value={option.title}>
                    {option.title}
                  </AutoComplete.Option>
                ))}
            </AutoComplete>
          </Col>
          <Col xs={{ order: 5, span: 2 }} sm={{ order: 3, span: 2 }}>
            <StatusIndicator loading={loading} warning={!!error?.length} />
          </Col>
          <Col flex="auto" xs={{ order: 2 }} sm={{ order: 4 }} />
          <Col xs={{ order: 3, span: 4 }} sm={{ order: 5, span: 2 }}>
            <Switch
              checkedChildren="°C"
              unCheckedChildren="°F"
              checked={this.state.isCelsius}
              onChange={this.onChangeUnit}
            />
          </Col>
        </Row>
        <Row>
          {forecastData &&
            forecastData.consolidated_weather.map((forecast) => (
              <Col xs={12} sm={6} md={4} key={forecast.id}>
                <WeatherItem
                  forecast={forecast}
                  isCelsius={this.state.isCelsius}
                />
              </Col>
            ))}
        </Row>
      </MainContent>
    );
  }
}

export const mapState = (state: any) => ({
  options: selectLocationSuggestion(state),
  forecastData: selectForecastData(state),
  loading: selectLoading(state),
  error: selectError(state),
});

export const mapDispatch = (dispatch: any) =>
  bindActionCreators(
    {
      getForecastByWoeId,
      getLocationSuggestions,
    },
    dispatch
  );

const withConnect = connect(mapState, mapDispatch);

export default withConnect(Weather);
