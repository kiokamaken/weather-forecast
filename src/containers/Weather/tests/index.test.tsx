import { AutoComplete, Select, Switch } from 'antd';
import { light } from 'components/Layout/Theme';
import StatusIndicator from 'components/StatusIndicator';
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import store from 'store';
import {
  getForecastByWoeId,
  getLocationSuggestions,
} from 'store/weather/slice';
import { ThemeProvider } from 'styled-components';

import Weather, { mapDispatch } from '..';
import { PageHeader } from '../Styled';

const fullMounted = () =>
  mount(
    <Provider store={store}>
      <ThemeProvider theme={light}>
        <Weather />
      </ThemeProvider>
    </Provider>
  );

describe('<Weather />', () => {
  const mockDispatch = jest.fn();

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('Should render without crashing', () => {
    const container = fullMounted();
    expect(container).toMatchSnapshot();
  });

  it('Should render PageHeader', () => {
    const container = fullMounted();
    expect(container.find(PageHeader).length).toEqual(1);
  });

  it('Should render two Select component', () => {
    const container = fullMounted();
    expect(container.find(Select).length).toEqual(2);
  });

  it('Should render Autocomplete component', () => {
    const container = fullMounted();
    expect(container.find(AutoComplete).length).toEqual(1);
  });

  it('Should render StatusIndicator component', () => {
    const container = fullMounted();
    expect(container.find(StatusIndicator).length).toEqual(1);
  });

  it('Should render Switch component', () => {
    const container = fullMounted();
    expect(container.find(Switch).length).toEqual(1);
  });

  it('Should update searchType value when change', () => {
    const container = fullMounted();
    const searchTypeSelect = container.find(Select).at(0);
    expect(searchTypeSelect.invoke('onChange')('lattlong')).toEqual('lattlong');
  });

  it('Should define getLocationSuggestions', () => {
    const result = mapDispatch(mockDispatch);
    expect(result.getLocationSuggestions).toBeDefined();
  });

  it('Should dispatch getLocationSuggestions when called', () => {
    const result = mapDispatch(mockDispatch);
    const payload = { search: 'abc', isLattLong: false };
    result.getLocationSuggestions(payload);
    expect(mockDispatch).toHaveBeenCalledWith(getLocationSuggestions(payload));
  });

  it('Should define getForecastByWoeId', () => {
    const result = mapDispatch(mockDispatch);
    expect(result.getForecastByWoeId).toBeDefined();
  });

  it('Should dispatch getForecastByWoeId when called', () => {
    const result = mapDispatch(mockDispatch);
    const payload = { id: '123' };
    result.getForecastByWoeId(payload);
    expect(mockDispatch).toHaveBeenCalledWith(getForecastByWoeId(payload));
  });
});
