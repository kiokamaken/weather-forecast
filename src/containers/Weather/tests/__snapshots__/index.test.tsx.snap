// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> Should render without crashing 1`] = `ReactWrapper {}`;
